framework/src/media/demux/mpeg2ts/TSDemuxer.cpp

```diff
@@ -0,0 +1,525 @@
+	meddbg("create TSDemuxer instance failed!\n");
+	return nullptr;
+}
+
+bool TSDemuxer::initialize(void)
+{
+	auto streamBuffer = stream::StreamBuffer::Builder()
+							.setBufferSize(CONFIG_DEMUX_BUFFER_SIZE)
+							.setThreshold(TS_DEMUX_BUFFER_THRESHOLD)
+							.build();
+	if (!streamBuffer) {
+		meddbg("streamBuffer is nullptr!\n");
+		return false;
+	}
+
+	mStreamBuffer = streamBuffer;
+	mBufferReader = std::make_shared<stream::StreamBufferReader>(mStreamBuffer);
+	mBufferWriter = std::make_shared<stream::StreamBufferWriter>(mStreamBuffer);
+
+	if (!mBufferReader || !mBufferWriter) {
+		meddbg("mBufferReader/Writer is nullptr!\n");
+		return false;
+	}
+
+	mParserManager = std::make_shared<ParserManager>();
+	if (!mParserManager) {
+		meddbg("mParserManager is nullptr!\n");
+		return false;
+	}
+
+	mPESParser = std::make_shared<PESParser>();
+	if (!mPESParser) {
+		meddbg("mPESParser is nullptr!\n");
+		return false;
+	}
+
+	mTSPacket = std::make_shared<TSPacket>();
+	if (!mTSPacket) {
+		meddbg("mTSPacket is nullptr!\n");
+		return false;
+	}
+
+	return true;
+}
+
+size_t TSDemuxer::getAvailSpace(void)
+{
+	return mBufferWriter->sizeOfSpace();
+}
+
+ssize_t TSDemuxer::pushData(uint8_t *buf, size_t size)
+{
+	size_t written = 0;
+	if (mBufferWriter) {
+		written += mBufferWriter->write(buf, size);
+	}
+	return (ssize_t)written;
+}
+
+ssize_t TSDemuxer::pullData(uint8_t *buf, size_t size, void *param)
+{
+	if (mPESPid == INVALID_PID) {
+		// setup PES pid
+		prog_num_t progNum;
+		if (param) {
+			progNum = *((prog_num_t *)param);
+		} else {
+			// use default 1st program
+			std::vector<prog_num_t> programs;
+			mParserManager->getPrograms(programs);
+			progNum = programs[0];
+		}
+		uint8_t streamType;
+		if (!mParserManager->getAudioStreamInfo(progNum, streamType, mPESPid)) {
+			meddbg("get audio PES PID failed\n");
+			return DEMUXER_ERROR_NOT_READY;
+		}
+		medvdbg("setup audio PES PID: 0x%x\n", mPESPid);
+	}
+
+	int ret = DEMUXER_ERROR_NONE;
+	size_t fill = 0;
+	size_t need;
+	while (fill < size) {
+		need = size - fill;
+		if (mPESParser->getESData() != nullptr) {
+			// get remaining payload in last PES packet
+			if (need > mPESParser->getESDataLen() - mPESDataUsed) {
+				need = mPESParser->getESDataLen() - mPESDataUsed;
+			}
+
+			memcpy(&buf[fill], mPESParser->getESData() + mPESDataUsed, need);
+			mPESDataUsed += need;
+			fill += need;
+			medvdbg("Got ES data %u(%u)/%u\n", fill, need, size);
+
+			if (mPESDataUsed == mPESParser->getESDataLen()) {
+				// all ES data in PES parser have been read.
+				medvdbg("All ES data (%u) in PES parser have been read!\n", mPESParser->getESDataLen());
+				mPESParser->reset();
+				mPESDataUsed = 0;
+			}
+			continue;
+		} else {
+			medvdbg("Need to get new PES packet!\n");
+		}
+
+		// get new PES packet
+		std::shared_ptr<PESPacket> pPESPacket = nullptr;
+		ret = getPESPacket(pPESPacket);
+		if (ret == DEMUXER_ERROR_WANT_DATA) {
+			medvdbg("Push more data to get PES packet\n");
```

Comment:
here you using verbose but below is error, what is correct?

Reply:
`DEMUXER_ERROR_WANT_DATA` is not critical error, actually it's normal case, so I used verbose log here.